onnxruntime/test/framework/execution_frame_test.cc

```diff
@@ -90,7 +90,7 @@ TEST_F(ExecutionFrameTest, TensorAllocationTest) {
   OrtValue* p_ml_value = frame.GetMutableNodeInputOrOutputMLValue(0);
   Tensor* p_tensor = p_ml_value ? p_ml_value->GetMutable<Tensor>() : nullptr;
   EXPECT_TRUE(p_tensor);
-  EXPECT_EQ(p_tensor->Shape(), shape);
+  EXPECT_EQ(*reinterpret_cast<const std::vector<int64_t>*>(&p_tensor->Shape()), *reinterpret_cast<const std::vector<int64_t>*>(&shape));
```

Comment:
`reinterpret_cast` assumes TensorShape and `std::vector<int64_t>` have the same layout, and will break if we add any member variables to TensorShape, right? Googling around I found that C styles cast work just fine here; they skip any access checks and you can cast to the private base class (I know raw C-style casts should be avoided, but I think it's better that reinterpret_cast in this case).

Reply:
No, it doesn't work. 
For the following code,
```c++
EXPECT_EQ((std::vector<int64_t>)(p_tensor->Shape()), *reinterpret_cast<const std::vector<int64_t>*>(&shape));
```
I got an error of:
```
In file included from /home/chasun/src/onnxruntime/cmake/external/googletest/googletest/include/gtest/gtest.h:360,
                 from /home/chasun/src/onnxruntime/onnxruntime/test/framework/execution_frame_test.cc:13:
/home/chasun/src/onnxruntime/onnxruntime/test/framework/execution_frame_test.cc: In member function ‘virtual void onnxruntime::test::ExecutionFrameTest_TensorAllocationTest_Test::TestBody()’:
/home/chasun/src/onnxruntime/onnxruntime/test/framework/execution_frame_test.cc:93:53: error: ‘std::vector<long int>’ is an inaccessible base of ‘onnxruntime::TensorShape’
   93 |   EXPECT_EQ((std::vector<int64_t>)(p_tensor->Shape()), *reinterpret_cast<const std::vector<int64_t>*>(&shape));

```

Reply:
> reinterpret_cast assumes TensorShape and std::vector<int64_t> have the same layout, and will break if we add any member variables to TensorShape, right?

Yes, that's the design. Please see the code in:
https://github.com/microsoft/onnxruntime/blob/master/include/onnxruntime/core/framework/tensor_shape.h#L132

The piece of code was there from very beginning, so the assumption must be always true, otherwise a lot of our code will be broken.

Reply:
You'd have to cast the pointer:

```
class A {};

class B : private A {};

B b;
A *ap = (A*)(&b);  // works
A *ap_static_cast = static_cast<A*>(&b);  // conversion to inaccessible base class A not allowed
A a = (A)(b);  // conversion to inaccessible base class A not allowed
```

But I'm fine with the current code since we have a static_assert for this assumption.